enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,391 @@
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +openshift:enable:FeatureSets=TechPreviewNoUpgrade
+type ImagePolicy struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ImagePolicySpec `json:"spec"`
+	// status contains the observed state of the resource.
+	// +optional
+	Status ImagePolicyStatus `json:"status,omitempty"`
+}
+
+// ImagePolicySpec is the specification of the ImagePolicy CRD.
+type ImagePolicySpec struct {
+	// scopes defines the list of image identities assigned to a policy. Each item refers to an image or repository in a registry implementing the "Docker Registry HTTP API V2".
+	// "scopes" uses one of the following:
+	// - complete image name, either using a tag or digest
+	// - prefixes of individual-image scopes
+	// - a wildcarded expression for matching all subdomains, the wildcard only presents at the beginning, *.example.com is a valid case, but example*.*.com is not.
+	// For more information about the format, see the document
+	// about the docker transport field:
+	// https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md#docker
+	// +kubebuilder:validation:Required
+	// +required
+	// +listType=set
+	Scopes []ImageScope `json:"scopes"`
+	// policy defines the verification policy for the items in the scopes list
+	// +kubebuilder:validation:Required
+	// +required
+	Policy Policy `json:"policy"`
+}
+
+// ImageScope is the item of the scopes list.
+// +kubebuilder:validation:Pattern=`^\*(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+$|^((?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(((?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+(?::[0-9]+)?)|(:[0-9]+))(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+:([\w][\w.-]{0,127}))?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}))?$`
+type ImageScope string
+
+// Policy defines the verification policy for the items in the scopes list.
+// +kubebuilder:validation:XValidation:rule="(has(self.rootOfTrust) && has(self.rootOfTrust.fulcioCAData)) ? has(self.fulcioSubject) : true",message="fulcioSubject must be set if fulcioCAData is set"
+// +kubebuilder:validation:XValidation:rule="has(self.fulcioSubject) ? has(self.fulcioSubject.oidcIssuer) && has(self.fulcioSubject.signedEmail): true",message="oidcIssuer and signedEmail must be set if fulcioSubject is set"
```

Comment:
Isn’t this covered by the `required` annotations on those fields?

(I have no idea, I don’t know the validation language.)

My (not very important) concern here is that it seems possible that we will add more fields / alternatives to `FulcioSubject`, and in that case it seems easier to maintain if the validation of the required fields / allowed field combinations only exists in one place, instead of two like it is now.

Reply:
I found the `required` does not work for the nested struct. the `required` does not appear in the auto-generated CRD yaml. 
I think we can keep the annotation here just to explain the field usage.